### PR TITLE
Async Locking

### PR DESCRIPTION
**Pull request by MelbourneDeveloper**
This PR introduces an async locking mechanism and makes it the default approach for dealing with singleton Futures.

**Problem**
The existing container does not lock on Futures. It's possible to start two+ Futures for the same type. This means that async work can run twice. Also, Futures that throw an error get stored and are permanently broken. This PR attempts to solve both these issues.

### CODE REVIEW

File: lib/ioc_container.dart [hunk 1]
```diff
@@ -254,37 +260,55 @@ extension IocContainerExtensions on IocContainer {
...
+
+      try {
+        //Await the locked call
+        final future = lock.execute();
+        await future;
+
+        //Store successful future
+        singletons[Future<T>] = future;
+
+        return future;
```
Comment (jogboms) [inline lib/ioc_container.dart]: Clever. I learnt something new with `await`ing on a `Future` that has already been `await`ed

File: lib/ioc_container.dart [hunk 2]
```diff
@@ -68,6 +73,9 @@ class IocContainer {
   ///so the container can store scope or singletons
   final Map<Type, Object> singletons;
 
+  // ignore: strict_raw_type, avoid_field_initializers_in_const_classes
+  final Map<Type, AsyncLock> locks;
```
Comment (jogboms) [inline lib/ioc_container.dart:77]: I noticed you don't use `pkg:meta` as a dependency I would have suggested `@visibleForTesting` in a bunch of places

**Review by jogboms — APPROVED** [review]
Looks good to me!

File: lib/ioc_container.dart (hunk 2 above)
Comment (MelbourneDeveloper) [reply lib/ioc_container.dart:77]: @jogboms I originally did use that, and at some point, I found a version library conflict with other packages between 1.7 and 1.8. At that point, I made a hard decision to remove all dependencies.

Also, exposing some of the internal workings allows you to access the singletons directly, for example, which was requested in an issue.

Comment (jogboms) [reply lib/ioc_container.dart:77]: Actually, you are the meta package is a plague when it comes to version conflicts.

**Comment by venbrinoDev** [comment]
This looks really good

**Comment by MelbourneDeveloper** [comment]
> This looks really good 

I have a feeling it will be

**Comment by MelbourneDeveloper** [comment]
> This looks really good 

Can you break it?

**Pull request merged**